File: packages/domain/src/query-packs/terms.ts
/**
 * Query term validation and public-safe projection.
 * Offensive historical terminology is never emitted as default public language.
 */
import type { PublicSafeTerm, QueryTerm, TermClass } from './types.js';
import { TERM_CLASSES } from './types.js';

export function isTermClass(value: string): value is TermClass {
  return (TERM_CLASSES as readonly string[]).includes(value);
}

export function assertTermClass(value: string): asserts value is TermClass {
  if (!isTermClass(value)) {
    throw new Error(`Unknown term class: ${value}`);
  }
}

export function assertQueryTermValid(term: QueryTerm): void {
  const text = term.text.trim();
  if (!text) {
    throw new Error('Query term text is required');
  }
  assertTermClass(term.termClass);
  if (term.termClass === 'source_specific' && !term.sourceId?.trim()) {
    throw new Error('source_specific terms require sourceId');
  }
  if (term.weight !== undefined && (term.weight < 0 || !Number.isFinite(term.weight))) {
    throw new Error('Query term weight must be a non-negative finite number');
  }
}

export function assertQueryTermsValid(terms: readonly QueryTerm[]): void {
  for (const term of terms) {
    assertQueryTermValid(term);
  }
}

function isResearchOnlyOffensive(term: QueryTerm): boolean {
  return term.researchOnlyOffensive === true;
}

/**
 * Strip or flag terms that must not appear in public interfaces.
 * Research-only offensive terms are omitted from public output entirely.
 */
export function toPublicSafeTerms(terms: readonly QueryTerm[]): readonly PublicSafeTerm[] {
  const safe: PublicSafeTerm[] = [];
  for (const term of terms) {
    if (isResearchOnlyOffensive(term)) {
      continue;
    }
    safe.push({
      text: term.text,
      termClass: term.termClass,
      redacted: false,
    });
  }
  return safe;
}

/**
 * Returns research terms including offensive historical language for internal query building.
 * Callers must never expose this output through public APIs without explicit review gates.
 */
export function toResearchQueryTerms(terms: readonly QueryTerm[]): readonly QueryTerm[] {
  return terms.map((term) => ({ ...term }));
}

export function countRedactedTerms(terms: readonly QueryTerm[]): number {
  return terms.filter((term) => isResearchOnlyOffensive(term)).length;
}

export function publicSafeSummary(terms: readonly QueryTerm[]): {
  readonly publicCount: number;
  readonly redactedCount: number;
} {
  const redactedCount = countRedactedTerms(terms);
  return {
    publicCount: terms.length - redactedCount,
    redactedCount,
  };
}

File: workers/research/src/black_book_research/query_packs/terms.py
"""Query term validation and public-safe projection (BB-038)."""

from __future__ import annotations

from .types import PublicSafeTerm, QueryTerm, TermClass

TERM_CLASSES: frozenset[TermClass] = frozenset(
    {
        "positive",
        "negative",
        "historical",
        "modern",
        "geographic",
        "alias",
        "source_specific",
    }
)


def assert_query_term_valid(term: QueryTerm) -> None:
    if not term.text.strip():
        raise ValueError("Query term text is required")
    if term.term_class not in TERM_CLASSES:
        raise ValueError(f"Unknown term class: {term.term_class}")
    if term.term_class == "source_specific" and not (term.source_id or "").strip():
        raise ValueError("source_specific terms require source_id")
    if term.weight is not None and (term.weight < 0 or term.weight != term.weight):
        raise ValueError("Query term weight must be a non-negative finite number")


def assert_query_terms_valid(terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> None:
    for term in terms:
        assert_query_term_valid(term)


def to_public_safe_terms(terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> tuple[PublicSafeTerm, ...]:
    safe: list[PublicSafeTerm] = []
    for term in terms:
        if term.research_only_offensive:
            continue
        safe.append(PublicSafeTerm(text=term.text, term_class=term.term_class, redacted=False))
    return tuple(safe)


def to_research_query_terms(terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> tuple[QueryTerm, ...]:
    return tuple(QueryTerm(**term.__dict__) for term in terms)


def count_redacted_terms(terms: tuple[QueryTerm, ...] | list[QueryTerm]) -> int:
    return sum(1 for term in terms if term.research_only_offensive)
